### ISSUE TITLE
Allow `useRedirect` to set location state

### ISSUE THREAD

**Issue by WiXSL** [marmelab/react-admin#6275]
Currently, the location state is hard-coded ([state](https://github.com/marmelab/react-admin/blob/adbd4e0d7b959b73fc9a9d4c27bb64e7928d70b9/packages/ra-core/src/sideEffect/useRedirect.ts#L58))
and not accessible from outside.

It would be nice to be able to pass state to the location, something like 

```js
const redirect = useRedirect();
const someState = { fromState: true};
redirect('edit', '/posts', 123, {}, someState);
```

**Comment by fzaninotto** [comment]
What's the use case for that?

**Comment by despatates** [comment]
I have a use case where I must prefill a `<Create />` form with some values ([doc](https://marmelab.com/react-admin/CreateEdit.html#prefilling-a-create-record)), and those values are depending on the result of a request to the server.
I can't ask my user to click on a button to redirect him, I want to do this automatically with a `redirect()`.

**Comment by LucusWebsites** [comment]
> What's the use case for that?

In same cases my resource edit page is accessible from different locations, so I would to return the user to the page where he came from.

So in my case I would to store the "fromUrl" in the state.

**Comment by fzaninotto** [comment]
Fixed by #6293
